monkey/common/di_container.py

```diff
@@ -0,0 +1,122 @@
+import inspect
+from typing import Any, MutableMapping, Sequence, Type, TypeVar
+
+T = TypeVar("T")
+
+
+class UnregisteredTypeError(ValueError):
+    pass
+
+
+class DIContainer:
+    """
+    A dependency injection (DI) container that uses type annotations to resolve and inject
+    dependencies.
+    """
+
+    def __init__(self):
+        self._type_registry = {}
+        self._instance_registry = {}
+
+    def register(self, interface: Type[T], concrete_type: Type[T]):
+        """
+        Register a concrete type that satisfies a given interface.
+
+        :param interface: An interface or abstract base class that other classes depend upon
+        :param concrete_type: A type (class) that implements `interface`
+        """
+        if not inspect.isclass(concrete_type):
+            raise TypeError(
+                "Expected a class, but received an instance of type "
+                f'"{concrete_type.__class__.__name__}"; Pass a class, not an instance, to '
+                "register(), or use register_instance() instead."
+            )
+
+        self._type_registry[interface] = concrete_type
+        DIContainer._del_key(self._instance_registry, interface)
+
+    def register_instance(self, interface: Type[T], instance: T):
+        """
+        Register a concrete instance that satisfies a given interface.
+
+        :param interface: An interface or abstract base class that other classes depend upon
+        :param instance: An instance (object) of a type that implements `interface`
+        """
+        self._instance_registry[interface] = instance
```

Comment:
We don't check if instance is of interface type because that's handled by typing?